# How does wildfire impact air quality?

## The goal of this notebook is to build multiple models exploring the impact of widlfire on air quality.



## 1. Multiple Linear Regression Model

### The goal of this section is to build a multiple linear regression model to predict air quality based on various predictors surrounding wildfire proximity.

#### y = ß0 + ß1X1 + ß2X2 + ... + ßnXn + error, where y is PM2.5 (fine particulate matter) levels, ß0 is a constant, X1 through Xn are the predictors, and ß1 through ßn are the predictors' coefficients.

### Imports

In [51]:
# Needed imports

import pandas as pd
import numpy as np
import os
import statsmodels.api as sm
import plotly.express as px
from sklearn.preprocessing import StandardScaler
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score

### Data Loading and Exploration

In [52]:
# Load the data
script_dir = os.getcwd()

df = pd.read_csv(f'{script_dir}/air_quality_weather_fires.csv')

df = df.dropna()

df

Unnamed: 0        date      site_id   latitude   longitude  state_name  \
0               0  2024-01-01  01-073-0023  33.553056  -86.815000     Alabama   
1               1  2024-01-01  04-013-9997  33.503833 -112.095767     Arizona   
2               2  2024-01-01  04-019-1028  32.295150 -110.982300     Arizona   
3               3  2024-01-01  05-119-0007  34.756189  -92.281296    Arkansas   
4               4  2024-01-01  06-001-0011  37.814781 -122.282347  California   
...           ...         ...          ...        ...         ...         ...   
19797       19797  2024-12-31  49-035-3015  40.777145 -111.945849        Utah   
19798       19798  2024-12-31  50-021-0002  43.608056  -72.982778     Vermont   
19799       19799  2024-12-31  51-087-0014  37.556520  -77.400270    Virginia   
19800       19800  2024-12-31  53-033-0080  47.568236 -122.308628  Washington   
19801       19801  2024-12-31  56-021-0100  41.182227 -104.778334     Wyoming   

      county_name           city_name  \
0       Jefferson          Birmingham   
1        Maricopa             Phoenix   
2            Pima              Tucson   
3         Pulaski   North Little Rock   
4         Alameda             Oakland   
...           ...                 ...   
19797   Salt Lake      Salt Lake City   
19798     Rutland             Rutland   
19799     Henrico  East Highland Park   
19800        King             Seattle   
19801     Laramie            Cheyenne   

                                         site_name   PM25  ...  \
0                                 North Birmingham  11.55  ...   
1                                    JLG SUPERSITE  85.35  ...   
2                            CHILDREN'S PARK NCore  16.30  ...   
3                                             PARR   5.90  ...   
4                                     Oakland West   6.90  ...   
...                                            ...    ...  ...   
19797                        Utah Technical Center   4.50  ...   
19798  State of Vermont District Court Parking Lot   4.70  ...   
19799                MathScience Innovation Center   5.60  ...   
19800                        SEATTLE - BEACON HILL   3.40  ...   
19801                               Cheyenne NCore   3.60  ...   

       fires_within_100km  has_nearby_fire    datetime  month  day_of_week  \
0                       3                1  2024-01-01      1            0   
1                       0                1  2024-01-01      1            0   
2                       0                1  2024-01-01      1            0   
3                       0                1  2024-01-01      1            0   
4                       9                1  2024-01-01      1            0   
...                   ...              ...         ...    ...          ...   
19797                   0                1  2024-12-31     12            1   
19798                   0                1  2024-12-31     12            1   
19799                   5                1  2024-12-31     12            1   
19800                   1                1  2024-12-31     12            1   
19801                   0                1  2024-12-31     12            1   

       is_weekend  season  wildfire_season  fire_distance_category  \
0               0  winter                0                   close   
1               0  winter                0                     far   
2               0  winter                0                     far   
3               0  winter                0                     far   
4               0  winter                0                   close   
...           ...     ...              ...                     ...   
19797           0  winter                0                     far   
19798           0  winter                0                     far   
19799           0  winter                0                   close   
19800           0  winter                0                moderate   
19801           0  winter                

In [53]:
df.T

0              1      \
Unnamed: 0                                  0              1   
date                               2024-01-01     2024-01-01   
site_id                           01-073-0023    04-013-9997   
latitude                            33.553056      33.503833   
longitude                             -86.815    -112.095767   
state_name                            Alabama        Arizona   
county_name                         Jefferson       Maricopa   
city_name                          Birmingham        Phoenix   
site_name                    North Birmingham  JLG SUPERSITE   
PM25                                    11.55          85.35   
CO                                        0.3            1.3   
O3                                      0.027          0.021   
NO2                                      24.3           37.3   
SO2                                       1.8            2.3   
AQI_PM25                                 57.0          176.0   
AQI_CO                                    3.0           15.0   
AQI_O3                                   25.0           19.0   
AQI_NO2                                    23             35   
AQI_SO2                                   1.0            3.0   
AQI                                      57.0          176.0   
temperature_2m_mean                       6.2           10.7   
temperature_2m_max                        9.9           16.5   
temperature_2m_min                        3.2            6.3   
relative_humidity_2m_mean                63.0           66.0   
wind_speed_10m_mean                      11.6            6.3   
wind_direction_10m_dominant             317.0           45.0   
precipitation_sum                         0.0            0.0   
precipitation_hours                       0.0            0.0   
et0_fao_evapotranspiration               1.78            1.4   
weather_code                         Overcast       Overcast   
distance_to_fire_km                 36.661749     114.379578   
fire_brightness                        298.82         296.16   
fire_frp                                 0.76           8.56   
fires_within_50km                           3              0   
fires_within_100km                          3              0   
has_nearby_fire                             1              1   
datetime                           2024-01-01     2024-01-01   
month                                       1              1   
day_of_week                                 0              0   
is_weekend                                  0              0   
season                                 winter         winter   
wildfire_season                             0              0   
fire_distance_category                  close            far   
fire_intensity                            low            low   

                                             2                  3      \
Unnamed: 0                                       2                  3   
date                                    2024-01-01         2024-01-01   
site_id                                04-019-1028        05-119-0007   
latitude                                  32.29515          34.756189   
longitude                                -110.9823         -92.281296   
state_name                                 Arizona           Arkansas   
county_name                                   Pima            Pulaski   
city_name                                   Tucson  North Little Rock   
site_name                    CHILDREN'S PARK NCore               PARR   
PM25                                          16.3                5.9   
CO                                             0.4                0.0   
O3                                           0.033              0.026   
NO2                                           22.2                4.5   
SO2                                            0.2                0.6   
AQI_PM25                                      67.0              

In [54]:
# Obtain a list of the columns for ease of typing
df.columns

Index(['Unnamed: 0', 'date', 'site_id', 'latitude', 'longitude', 'state_name',
       'county_name', 'city_name', 'site_name', 'PM25', 'CO', 'O3', 'NO2',
       'SO2', 'AQI_PM25', 'AQI_CO', 'AQI_O3', 'AQI_NO2', 'AQI_SO2', 'AQI',
       'temperature_2m_mean', 'temperature_2m_max', 'temperature_2m_min',
       'relative_humidity_2m_mean', 'wind_speed_10m_mean',
       'wind_direction_10m_dominant', 'precipitation_sum',
       'precipitation_hours', 'et0_fao_evapotranspiration', 'weather_code',
       'distance_to_fire_km', 'fire_brightness', 'fire_frp',
       'fires_within_50km', 'fires_within_100km', 'has_nearby_fire',
       'datetime', 'month', 'day_of_week', 'is_weekend', 'season',
       'wildfire_season', 'fire_distance_category', 'fire_intensity'],
      dtype='object')

### Variable Selection

In [55]:
# List numeric variables
# Only use mean temperature (instead of 'temperature_2m_max', 'temperature_2m_min'), mean wind speed (instead of 'wind_direction_10m_dominant'), and precipitation sum (instead of precipitation_hours') to avoid multicollinearity
# Temperature, humidity, and evaporatranspiration are often correlated, so remove the latter two variables
X_num = ['latitude', 'longitude', 'temperature_2m_mean', 'wind_speed_10m_mean', 'precipitation_sum', 'fires_within_50km', 'fires_within_100km', 'distance_to_fire_km', 'fire_brightness', 'fire_frp']

# Convert X_num to numerics, not strings
df[X_num] = df[X_num].apply(pd.to_numeric, errors='coerce')
df = df.dropna(subset=X_num + ['PM25']).reset_index(drop=True)

# Has nearby fire and wildfire season are already binary
# However, has nearby fire is described by the number of fires within 50 and 100 km - remove to avoid multicollinearity 
X_binary = ['wildfire_season']

# List categorical variables
# Site_name, Site_ID, state_name, county_name, city_name, and site name are multicollinear with lat and lon
# Fire distance category is correlated with the distance to fire in km so remove that 
X_cat = ['weather_code', 'season', 'fire_intensity']

# Time data are not used as predictors in our model; do not include those

# Reserve y variable in a separate df
y_df = df[['PM25']].copy() 

# Select binary and categorical X variables (still need to scale numeric variables)
selected_columns = X_binary + X_cat 
X_df = df[selected_columns].reset_index(drop=True)

# Scale our numeric variables
scaler = StandardScaler()
scaled_num_X = scaler.fit_transform(df[X_num])
scaled_X_df = pd.DataFrame(scaled_num_X, columns=X_num)

# Combine scaled_X_df and X_df
X_df = pd.concat([X_df, scaled_X_df], axis=1)

# Use One-Hot_Encoding for categorical variables
X_df_processed = pd.get_dummies(X_df, 
                                 columns=X_cat, 
                                 prefix='dummy', 
                                 drop_first=True)


In [56]:
cols = X_df_processed.columns.tolist()

In [57]:
# Set up X and y 
X = X_df_processed[cols].astype('float64')
X = sm.add_constant(X)
y = y_df[['PM25']].to_numpy(dtype='float64')

In [58]:
# Check for multicollinearity by analyzing the Variance Inflation Factor (VIF)
def vif(X):
    vif_data = pd.DataFrame()
    vif_data["feature"] = X.columns

    vif_data["VIF"] = [variance_inflation_factor(X.values, i)
                            for i in range(len(X.columns))]
    print(vif_data)

vif(X)

                     feature          VIF
0                      const  5154.975771
1            wildfire_season     5.643403
2                   latitude     2.222666
3                  longitude     1.316024
4        temperature_2m_mean     3.845416
5        wind_speed_10m_mean     1.165285
6          precipitation_sum     3.147396
7          fires_within_50km     1.771329
8         fires_within_100km     1.801214
9        distance_to_fire_km     1.329948
10           fire_brightness     1.555408
11                  fire_frp     6.763058
12                   dummy_1     1.700684
13                   dummy_2     1.853335
14                   dummy_3     5.536279
15                  dummy_51     3.145192
16                  dummy_53     2.069827
17                  dummy_55     1.581021
18                  dummy_61     2.198105
19                  dummy_63     2.232277
20                  dummy_65     1.226222
21                  dummy_71     1.614577
22                  dummy_73     1

In [59]:
# Remove variables with moderate to strong multicollinearity (above 5) to improve model performance

# Fire frp and fire brightness instinctively will be correlated, so remove fire frp as per its VIF above 5
# Wildfire season has a high VIF - likely correlated to season
# The category fire_intensity has high VIF - likely due to multicollinearity with fire brightness - remove

# Important note: high VIFs within a single categorical variable are not necessarily a concern
# They likely indicate a small number of reference category cases due to our relatively small dataset
# We will address this later with bootstrapping
# One change to make is to combine certain weather_code values to increase the number of values in a category (will reduce VIF)


"""
Combine cells into clear, cloudy, rainy, snowy:

Notes:
1 ("Clouds generally dissolving or becoming less developed")
2 ("State of sky on the whole unchanged")
3 ("Clouds generally forming or developing")
51 ("Drizzle, not freezing, continuous - slight at time of observation")
53 ("Drizzle, not freezing, continuous - moderate at time of observation")
55 ("Drizzle, not freezing, continuous - heavy (dense) at time of observation") 
61 ("Rain, not freezing, continuous - slight at time of observation")
63 ("Rain, not freezing, continuous - moderate at time of observation"))
65 ("Rain, not freezing, continuous - heavy at time of observation")  
71 ("Continuous fall of snowflakes - slight at time of observation")
73 ("Continuous fall of snowflakes - moderate at time of observation")
75 ("Continuous fall of snowflakes - heavy at time of observation")

So:

Clear => 1, Clear sky, Mainly clear      
2 => Does not tell us about the actual state of the weather, remove
Cloudy => 3, Overcast, Partly cloudy 
Rainy => 51, 53, 55, 61, 63, 65, Dense drizzle, Heavy rain, Light drizzle, Moderate drizzle, Moderate rain, Slight rain
Snowy => 71, 73, 75, Heavy snow fall, Moderate snow fall, Slight snow fall 
"""

df['weather_code'] = df['weather_code'].replace(['1', 'Clear sky', 'Mainly clear'], 'clear')
df = df[df.weather_code != '2']
df['weather_code'] = df['weather_code'].replace(['3', 'Overcast', 'Partly cloudy'], 'cloudy')
df['weather_code'] = df['weather_code'].replace(['51', '53', '55', '61', '63', '65', 'Dense drizzle', 'Heavy rain', 'Light drizzle', 'Moderate drizzle', 'Moderate rain', 'Slight rain'], 'rainy')
df['weather_code'] = df['weather_code'].replace(['71', '73', '75', 'Heavy snow fall', 'Moderate snow fall', 'Slight snow fall'], 'snowy')

# List numeric variables
X_num = ['latitude', 'longitude', 'temperature_2m_mean', 'wind_speed_10m_mean', 'precipitation_sum', 'fires_within_50km', 'fires_within_100km', 'distance_to_fire_km', 'fire_brightness']

# Convert X_num to numerics, not strings
df[X_num] = df[X_num].apply(pd.to_numeric, errors='coerce')

# List categorical variables
# Site_name, Site_ID, state_name, county_name, city_name, and site name are multicollinear with lat and lon
# Fire distance category is correlated with the distance to fire in km so remove that 
X_cat = ['weather_code', 'season']

# Need to drop NaN values again because of the error coercion in X_num
# Drop NaNs only from columns we actually care about
# We include y ('PM25') in the subset to ensure we have targets for all rows
cols_to_check = X_num + X_cat + ['PM25']
df = df.dropna(subset=cols_to_check)

# Reserve y variable in a separate df
y_df = df[['PM25']].copy().reset_index(drop=True)

# Select categorical X variables (still need to scale numeric variables)
X_df_cat = df[X_cat].reset_index(drop=True)

# Select the numeric data from the X_df to scale
X_num_data = df[X_num]

# Scale our numeric variables
scaler = StandardScaler()
scaled_num_X = scaler.fit_transform(X_num_data)
scaled_X_df = pd.DataFrame(scaled_num_X, columns=X_num).reset_index(drop=True)

# Combine scaled_X_df and X_df
X_df = pd.concat([X_df_cat, scaled_X_df], axis=1)

# Use One-Hot_Encoding for categorical variables
X_df_processed = pd.get_dummies(X_df, 
                                columns=X_cat, 
                                prefix='dummy', 
                                drop_first=True)

cols = X_df_processed.columns.tolist()

# Set up X and y 
X = X_df_processed[cols].astype('float64')
X = sm.add_constant(X)
y = y_df[['PM25']].to_numpy(dtype='float64')

In [60]:
# Retest multicollinearity with VIF
vif_data = pd.DataFrame()
vif_data["feature"] = X.columns

vif_data["VIF"] = [variance_inflation_factor(X.values, i)
                          for i in range(len(X.columns))]
print(vif_data)

                feature        VIF
0                 const  74.749165
1              latitude   1.555117
2             longitude   1.247931
3   temperature_2m_mean   3.328919
4   wind_speed_10m_mean   1.155813
5     precipitation_sum   1.247420
6     fires_within_50km   1.770387
7    fires_within_100km   1.796615
8   distance_to_fire_km   1.285613
9       fire_brightness   1.012701
10          dummy_clear   9.181738
11         dummy_cloudy  17.865094
12          dummy_rainy  18.454052
13          dummy_snowy   4.952666
14         dummy_spring   1.634281
15         dummy_summer   2.012294
16         dummy_winter   2.311232


In [61]:
# Now we can see that weather has high VIF regardless of the category, so remove this
# This are likely multicollinear with season (spring, summer, winter)

# List numeric variables as X, since only variables left are numeric
X_cols = ['latitude', 'longitude', 'temperature_2m_mean', 'wind_speed_10m_mean', 'precipitation_sum', 'fires_within_50km', 'fires_within_100km', 'distance_to_fire_km', 'fire_brightness']

# Convert X_num to numerics, not strings
df[X_cols] = df[X_cols].apply(pd.to_numeric, errors='coerce')

# Drop rows with NaN values after conversion and row removal in previous steps/ This aligns X and y within the main 'df' DataFrame.
df.dropna(subset=['PM25'] + X_cols, inplace=True)

# Reserve y variable in a separate df and reset index
y_df = df[['PM25']].copy().reset_index(drop=True)

# Select appropriate X variables and name it all such that we will scale it 
X_data = df[X_cols]

# Scale our  variables
scaler = StandardScaler()
scaled_X = scaler.fit_transform(X_data)
scaled_df = pd.DataFrame(scaled_X, columns=X_cols)

# Set up X and y 
X = sm.add_constant(scaled_df)
y = y_df[['PM25']].to_numpy(dtype='float64')

In [62]:
# Check VIF one last time
vif_data = pd.DataFrame()
vif_data["feature"] = X.columns

vif_data["VIF"] = [variance_inflation_factor(X.values, i)
                          for i in range(len(X.columns))]
print(vif_data)

               feature       VIF
0                const  1.000000
1             latitude  1.320852
2            longitude  1.206133
3  temperature_2m_mean  1.228414
4  wind_speed_10m_mean  1.130307
5    precipitation_sum  1.070557
6    fires_within_50km  1.769330
7   fires_within_100km  1.782481
8  distance_to_fire_km  1.246692
9      fire_brightness  1.009473


### Modeling and  Bootstrapping

In [63]:
# Make X and y pd dfs
X = pd.DataFrame(X)
y = pd.DataFrame(y)
y = y.rename(columns={0:'PM25'})

In [64]:
# Make a baseline model
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size = 0.30, random_state = 42) 

X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size = 0.50, random_state = 42) 

model1 = sm.OLS(y_train, X_train).fit()

print(model1.summary())

                            OLS Regression Results                            
Dep. Variable:                   PM25   R-squared:                       0.127
Model:                            OLS   Adj. R-squared:                  0.126
Method:                 Least Squares   F-statistic:                     220.1
Date:                Thu, 11 Dec 2025   Prob (F-statistic):               0.00
Time:                        13:36:17   Log-Likelihood:                -40694.
No. Observations:               13659   AIC:                         8.141e+04
Df Residuals:                   13649   BIC:                         8.148e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   7.4029    

In [65]:
# Bootstrap entire process of making a model

def boot(X, y, S):
    
    # Create an array to store the coefficients for each model
    coef_store = np.zeros((S, X.shape[1]))

    n = len(X)

    # Divide each dataset into a test/train split
    for dataset in range(S):

        # Sample indices with replacement
        idx = np.random.choice(n, n, replace=True)

        # Bootstrap samples
        X_b = X.iloc[idx]
        y_b = y.iloc[idx]

        # Fit the model on the training data
        boot_model1 = sm.OLS(y_b,X_b).fit()

        # Store the coefficients
        coef_store[dataset, :] = boot_model1.params

    # Store results as a dataframe
    boot_results = pd.DataFrame(coef_store, columns=X.columns)

    return boot_results.mean(), boot_results.std()

# Bootstrap
mean, std = boot(X_train, y_train, 1000)

print(f"\nBootstrap Coefficient Means: \n{mean} \n\nBootstrap Coefficient Standard Errors: \n{std}")


Bootstrap Coefficient Means: 
const                  7.404964
latitude              -0.696399
longitude              0.193699
temperature_2m_mean    0.405655
wind_speed_10m_mean   -1.069895
precipitation_sum     -0.471953
fires_within_50km     -0.056751
fires_within_100km     0.437386
distance_to_fire_km   -0.472029
fire_brightness       -0.124867
dtype: float64 

Bootstrap Coefficient Standard Errors: 
const                  0.040250
latitude               0.049900
longitude              0.043007
temperature_2m_mean    0.048191
wind_speed_10m_mean    0.045449
precipitation_sum      0.030768
fires_within_50km      0.219097
fires_within_100km     0.146885
distance_to_fire_km    0.035435
fire_brightness        0.040067
dtype: float64


In [66]:
# Compare the bootstrapped coefficients' standard errors to the OLS standard errors
# If they are similar: OLS assumptions are met, traditional inference is reliable
# If they differ greatly: OLS assumptions may be violated; trust bootstrap SEs for inference
# Large standard errors (regardless of method) suggest high uncertainty in coefficient estimates

# OLS standard errors (from summary)
ols_se = model1.bse

# Compare them
comparison = pd.DataFrame({
    'Bootstrap SE': std,
    'OLS SE': ols_se,
    'Abs Val of Diff': np.abs(std - ols_se)
})

print(comparison)

                     Bootstrap SE    OLS SE  Abs Val of Diff
const                    0.040250  0.040750         0.000500
latitude                 0.049900  0.046630         0.003270
longitude                0.043007  0.044741         0.001733
temperature_2m_mean      0.048191  0.045057         0.003134
wind_speed_10m_mean      0.045449  0.043167         0.002282
precipitation_sum        0.030768  0.042380         0.011612
fires_within_50km        0.219097  0.047194         0.171904
fires_within_100km       0.146885  0.048476         0.098409
distance_to_fire_km      0.035435  0.045291         0.009856
fire_brightness          0.040067  0.040994         0.000927


They are similar: OLS assumptions are met, traditional inference is reliable

### Residual Examination

In [67]:
df["fitted_vals"]=model1.fittedvalues
df["residuals"]=model1.resid

In [68]:
fig=px.scatter(df, x="fitted_vals", y="residuals")
fig.add_hline(y=0, line_dash="dash", line_color="red")
fig.show() 

As our fitted values for PM25 increase, the spread of residuals increases.

More of our residuals are positive, and some rise as high as almost 80.

This, combined with the low R-squared and adjusted R-squared, leads us to conclude that our model is not very accurate at predicting, despite the fact that all predictors were highly significant

### Comparing Train and Validate Set

In [69]:
def rmse(y_true, y_pred):
    return mean_squared_error(y_true, y_pred) ** 0.5

rows = []

yhat_tr = model1.predict(X_train)
yhat_val = model1.predict(X_test)

rows.append({
    "Train RMSE": rmse(y_train, yhat_tr),
    "Validate RMSE": rmse(y_test, yhat_val),
    "Train R2": r2_score(y_train, yhat_tr),
    "Validate R2": r2_score(y_test, yhat_val),
})

results = pd.DataFrame(rows).sort_values("Validate RMSE")
print(results.round(4))

   Train RMSE  Validate RMSE  Train R2  Validate R2
0      4.7604         4.7662    0.1268       0.1059


Our train and validate RMSE and R-squared values do not differ significantly. This is a sign that our model adapts well to new data and is robust.

### Ridge Regularization

In [70]:
# Create Ridge regression models with various alphas to find the best one
alphas = [0.01, 0.1, 1.0, 10.0, 100.0]
best_alpha = None
best_val_mse = float('inf')

for alpha in alphas:
    # Train on training set
    ridge_model = Ridge(alpha=alpha)
    ridge_model.fit(X_train, y_train)
    
    # Evaluate on validation set
    y_pred_val = ridge_model.predict(X_val)
    val_mse = mean_squared_error(y_val, y_pred_val)

    print(f"Alpha: {alpha}, Validation MSE: {val_mse}")
    
    if val_mse < best_val_mse:
        best_val_mse = val_mse
        best_alpha = alpha

print(f"\nBest alpha: {best_alpha}")

Alpha: 0.01, Validation MSE: 25.71901851866972
Alpha: 0.1, Validation MSE: 25.719019018112803
Alpha: 1.0, Validation MSE: 25.71902402307475
Alpha: 10.0, Validation MSE: 25.719075124028365
Alpha: 100.0, Validation MSE: 25.719689515858636

Best alpha: 0.01


A note that the validation MSEs for all alphas are very similar. Still use the best one, of course.

In [71]:
# Evaluate on validate set
final_val_model = Ridge(alpha=best_alpha)
final_val_model.fit(X_val, y_val)

y_pred_val = final_val_model.predict(X_val)
val_mse = mean_squared_error(y_val, y_pred_val)

print(f"Final Validate MSE: {val_mse}")
print(f"Final Validate RMSE: {np.sqrt(val_mse)}")

Final Validate MSE: 25.456665867187606
Final Validate RMSE: 5.045459926229482


The ridge model performs worse than the OLS model without ridge, as it has a higher RMSE value (5.0455) than the OLS validate RMSE (4.7662). Try lasso next.

### Lasso Regularization

In [73]:
# Create Lasso  models with various alphas to find the best one
alphas = [0.01, 0.1, 1.0, 10.0, 100.0]
best_alpha = None
best_val_mse = float('inf')

# Use the train data to find the best alpha
for alpha in alphas:
    # Train on training set
    lasso_model = Lasso(alpha=alpha, fit_intercept=True) 
    lasso_model.fit(X_train, y_train)
    
    # Evaluate on validation set
    y_pred_val = lasso_model.predict(X_val)
    val_mse = mean_squared_error(y_val, y_pred_val)

    print(f"Alpha: {alpha}, Validation MSE: {val_mse}")
    
    if val_mse < best_val_mse:
        best_val_mse = val_mse
        best_alpha = alpha

print(f"\nBest alpha: {best_alpha}")

Alpha: 0.01, Validation MSE: 25.718779204565983
Alpha: 0.1, Validation MSE: 25.804375474965898
Alpha: 1.0, Validation MSE: 28.227853332962603
Alpha: 10.0, Validation MSE: 29.082901061440513
Alpha: 100.0, Validation MSE: 29.082901061440513

Best alpha: 0.01


In [74]:
# Evaluate on validate set
final_val_model = Lasso(alpha=best_alpha)
final_val_model.fit(X_val, y_val)

y_pred_val = final_val_model.predict(X_val)
val_mse = mean_squared_error(y_val, y_pred_val)

print(f"Final Validate MSE: {val_mse}")
print(f"Final Validate RMSE: {np.sqrt(val_mse)}")

Final Validate MSE: 25.45758088128955
Final Validate RMSE: 5.045550602391136


The lasso model also performs worse than the OLS model without ridge, as it has a higher RMSE value (5.0456) than the OLS validate RMSE (4.7662). Stick with OLS.

### Final Model Testing - OLS (No Ridge or Lasso)

In [ ]:
model2 = sm.OLS(y_test,X_test).fit()

print(model2.summary())

                            OLS Regression Results                            
Dep. Variable:                   PM25   R-squared:                       0.112
Model:                            OLS   Adj. R-squared:                  0.109
Method:                 Least Squares   F-statistic:                     40.78
Date:                Thu, 11 Dec 2025   Prob (F-statistic):           4.22e-69
Time:                        13:26:13   Log-Likelihood:                -8717.2
No. Observations:                2928   AIC:                         1.745e+04
Df Residuals:                    2918   BIC:                         1.751e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   7.4990    

In [ ]:
def rmse(y_true, y_pred):
    return mean_squared_error(y_true, y_pred) ** 0.5

rows = []

yhat_test = model2.predict(X_test)

rows.append({
    "Test RMSE": rmse(y_test, yhat_test),
    "Test R2": r2_score(y_test, yhat_test),
})

results = pd.DataFrame(rows).sort_values("Test RMSE")
print(results.round(4))

   Test RMSE  Test R2
0     4.7505   0.1117


In [ ]:
df["fitted_vals"]=model2.fittedvalues
df["residuals"]=model2.resid

fig=px.scatter(df, x="fitted_vals", y="residuals")
fig.add_hline(y=0, line_dash="dash", line_color="red")
fig.show() 

### Conclusions

The residual plot for the model fitted to the validated model, as well as the RMSE and R-squared values, are similar to what was seen on the training and validation data.

The residual plot has the same flaws as previously noted, and our R-squared is low. These indicate a poor model fit.

To interpret the RMSE, we must determine the scaled range of the PM25 variable.

In [ ]:
np.max(y_val['PM25']) - np.min(y_val['PM25'])

88.4

An RMSE of 4.7505 on a scale of almost 50 means that our predictions are off by about 8% of the range on average (about 4.7505/49.2). This means that our predictions are usually reasonable, despite being not very accurate, as indicated by our low R-squared. We conclude middling results for this model.